[Federation] Downsize the release binary distribution v2.

madhusudancs opened the pull request:
Second attempt of PR #29632.

There are two things that this PR does:
1. It removes `federation-apiserver` and `federation-controller-manager` from binaries and docker_wrapped_binaries target lists.
2. Build the docker image for `hyperkube` on-the-fly while pushing the federation images. 

``` release-note
Federation binaries and their corresponding docker images - `federation-apiserver` and `federation-controller-manager` are now folded in to the `hyperkube` binary. If you were using one of these binaries or docker images, please switch to using the `hyperkube` version. Please refer to the federation manifests - `federation/manifests/federation-apiserver.yaml` and `federation/manifests/federation-controller-manager-deployment.yaml` for examples.
```

cc @kubernetes/sig-cluster-federation @colhom 

Fixes Issue #28633

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29929)

<!-- Reviewable:end -->

madhusudancs committed:
[Federation] Do not build separate binaries for federation control plane components, hyperkube should be sufficient.

Also build the hyperkube docker image on-the-fly.

This is only a temporary fix until the proposal in issue
https://github.com/kubernetes/kubernetes/issues/28630 is implemented.

Also, the new build/deployment method completely obviates this step.

We use debian image instead of busybox and do not build hyperkube as a
static binary yet. Wait until PR
https://github.com/kubernetes/kubernetes/pull/26028 is merged to build
static hyperkube binaries.

madhusudancs committed:
Limit sourcing the build scripts only for build/push phase of federation control plane components.

david-mcmahon commented:
Reminder to self, update currently in-process move of some of this change to [kubernetes/release](https://github.com/kubernetes/release/blob/master/lib/releaselib.sh#L826) when this goes in.

luxas commented:
The big blocker here is: https://github.com/kubernetes/kubernetes/pull/26028

hyperkube is now dynamically linked in order to include the kubelet, which makes it unrunnable in a plain busybox container. Also, hyperkube shouldn't be in the `STATIC_BINARIES` list.

I have some small other comments also, but we'll have to wait for #26028 (which I'd like to get in ASAP)

madhusudancs commented:
@luxas what's the ETA on #26028. Is it in O(days), O(weeks), ...? Can we wrap the hyperkube binary in an ubuntu image to unblock this PR?

luxas commented:
ETA is days (today) for me, but the node e2e tests are haunting me :(
`busybox:glibc` exists that may unblock, but be sure to build for all platforms (just add `:glibc` at the end)!

madhusudancs commented:
@luxas I can definitely wait if it is today or if it is in the next couple of days. Could you please ping this PR once #26028 is merged?

luxas commented:
I will do, hopefully it's soon

madhusudancs commented:
@luxas Ping, since 1.4 code freeze is Monday.

luxas commented on federation/cluster/common.sh:
```diff
@@ -224,58 +223,68 @@ function create-federation-api-objects {
-	echo "Tag: ${srcImageName} --> ${dstImageName}"
-	docker tag -f "$srcImageName" "$dstImageName"
+        ln "${bin_path}" "${docker_build_path}/${binary}"
+        printf " FROM busybox \n ADD ${binary} /usr/local/bin/${binary}\n" > ${docker_file_path}
```
`FROM busybox:glibc` since hyperkube is dynamically linked.
Also, this should be built for all architectures: (IIRC)
 - armel/busybox:glibc
 - aarch64/busybox:glibc
 - ppc64le/busybox:glibc

luxas commented on hack/lib/golang.sh:
```diff
@@ -169,8 +167,7 @@ readonly KUBE_STATIC_LIBRARIES=(
   kubectl
-  federation-apiserver
-  federation-controller-manager
+  hyperkube
```
This is wrong, hyperkube is dynamically linked because of kubelet (until I get my PR working)

luxas commented:
@madhusudancs Reviewed. My PR hangs in strange ways on node e2e gci and coreos, haven't fully figured out why. So it won't probably make v1.4.

We can use the alternate approach with busybox + glibc instead.

madhusudancs commented:
@luxas addressed the comments PTAL.

---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/cluster/common.sh, line 258 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29929#-KOh0VrkjljOpJCIk_xP-r1-258:-KPzZhtGNlWRHWVbkeAT:b-om16hf) ([raw file](https://github.com/kubernetes/kubernetes/blob/26962d07be8e517acf1a304d3aa4d2e82bf4da46/federation/cluster/common.sh#L258)):*
<details><summary><i>Previously, luxas (Lucas Käldström) wrote…</i></summary>
> `FROM busybox:glibc` since hyperkube is dynamically linked.
> 
> Also, this should be built for all architectures: (IIRC)
> 
>  - armel/busybox:glibc
> 
>  - aarch64/busybox:glibc
> 
>  - ppc64le/busybox:glibc
</details>
This code path is used only during development. Let's please get amd64 out of the way first, so that it can be used and/tested soon? I will be happy to add support for other architectures as a follow up if necessary. Also, I am actively working on obsoleting these scripts, so there is some conflict of interest in perfecting this scripts.

---

*[hack/lib/golang.sh, line 170 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29929#-KOh0VrkjljOpJCIk_xS-r1-170:-KPz_QtsU5c4EbnICOXO:bss4a2b) ([raw file](https://github.com/kubernetes/kubernetes/blob/26962d07be8e517acf1a304d3aa4d2e82bf4da46/hack/lib/golang.sh#L170)):*
<details><summary><i>Previously, luxas (Lucas Käldström) wrote…</i></summary>
> This is wrong, hyperkube is dynamically linked because of kubelet (until I get my PR working)
</details>
Removed.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29929#-:-KPzawahg_3AzO3T10jc:b-5rr0l3)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Reviewed 1 of 5 files at r2.
Review status: 1 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/cluster/common.sh, line 258 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29929#-KOh0VrkjljOpJCIk_xP-r1-258:-KQ-WnEwq9-F8JxcXk8Y:breui8g) ([raw file](https://github.com/kubernetes/kubernetes/blob/26962d07be8e517acf1a304d3aa4d2e82bf4da46/federation/cluster/common.sh#L258)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> This code path is used only during development. Let's please get amd64 out of the way first, so that it can be used and/tested soon? I will be happy to add support for other architectures as a follow up if necessary. Also, I am actively working on obsoleting these scripts, so there is some conflict of interest in perfecting this scripts.
</details>
@luxas btw, I am giving up on busybox and going with ubuntu. There are far more libraries missing than glibc. `hyperkube` also needs libpthread for example.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29929)*
<!-- Sent from Reviewable.io -->

luxas commented on federation/cluster/common.sh:
```diff
@@ -267,58 +266,68 @@ function create-federation-apiserver-certs {
-	local dstImageTag="${FEDERATION_IMAGE_TAG:-$srcImageTag}"
-	local srcImageName="${FEDERATION_IMAGE_REPO_BASE}/${binary}:${srcImageTag}"
-	local dstImageName="${FEDERATION_PUSH_REPO_BASE}/${binary}:${dstImageTag}"
+        rm -rf ${docker_build_path}
```
There should be a loop here for all values in `KUBE_SERVER_PLATFORMS`, and the correct base image should be chosen.

Please use debian. You may see the base image values for all arches in the hyperkube Makefile

luxas replied:
@madhusudancs There should be a loop here for all values in KUBE_SERVER_PLATFORMS, and the correct base image should be chosen.

like in `build/common.sh`
```bash
for arch in "${KUBE_SERVER_PLATFORMS[@]##*/}"; do
  kube::log::status "Building hyperkube image for arch: ${arch}"
  REGISTRY="${KUBE_DOCKER_REGISTRY}" VERSION="${KUBE_DOCKER_IMAGE_TAG}" ARCH="${arch}" make -C cluster/images/hyperkube/ build
done
```

madhusudancs replied:
@luxas it doesn't make sense to add that loop here. These images are built on the fly,  only during development. It doesn't make sense to build the images for all the platforms while you are deploying for one of them. 

One thing we could do is build for $KUBE_ARCH. But I will do that in a follow up.

luxas replied:
Okay, didn't get that this was only for development.

How about the release then? I assume these images should be released, and we have to have code for that also

madhusudancs replied:
We don't need special code for the release. We push the hyperkube images to gcr.io/google_containers.

madhusudancs commented:
@luxas PTAL.

---

Reviewed 2 of 5 files at r2, 2 of 2 files at r5.
Review status: 2 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[federation/cluster/common.sh, line 297 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29929#-KOh0VrkjljOpJCIk_xP-r5-297:-KQ21c3K4g3fSbg9RnnY:bmycjru) ([raw file](https://github.com/kubernetes/kubernetes/blob/83c4124717ffa3166a155fd330d02e866534e27e/federation/cluster/common.sh#L297)):*
> There should be a loop here for all values in KUBE_SERVER_PLATFORMS, and the correct base image should be chosen.

Please see the comment below.

> Please use debian. You may see the base image values for all arches in the hyperkube Makefile

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29929#-:-KQ21i6KcXXcFnpFN7Ou:b-p3m14r)*
<!-- Sent from Reviewable.io -->

luxas commented:
cc @david-mcmahon PTAL as well

@madhusudancs Please squash your commits

madhusudancs commented:
@luxas folded the last commit into one of the previous commit. I would like to have the other commits separate for posterity.

luxas commented:
LGTM

(Haven't had time to test it manually, but if you say it works, it's okay for me)

Also defer to @david-mcmahon, which should apply the label if he thinks it's okay
@nikhiljindal Any concerns or does it look good to you?
